File: Java/_113_Path_Sum_II/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> curr = new ArrayList<Integer>();
        dfs(root, sum, res, curr);
        return res;
    }
    
    public void dfs(TreeNode root, int sum, List<List<Integer>> res, List<Integer> curr) {
        if (root == null)
            return;
        if (root.val == sum && root.right == null && root.left == null) {
            curr.add(root.val);
            res.add(curr);
            return;
        }
        curr.add(root.val);
        dfs(root.left, sum - root.val, res, new ArrayList<Integer>(curr));
        dfs(root.right, sum - root.val, res, new ArrayList<Integer>(curr));
    }
}

File: Python/_113_Path_Sum_II/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        def helper(root, sum, res, curr):
            if not root:
                return
            if root.val == sum and (not root.right) and (not root.left):
                curr.append(root.val)
                res.append(curr)
                return
            helper(root.left, sum - root.val, res, curr + [root.val])
            helper(root.right, sum - root.val, res, curr + [root.val])
        res = []
        helper(root, sum, res, [])
        return res
